gdal/ogr/ogrfeature.cpp

```diff
@@ -2036,14 +2036,54 @@ static void OGRFeatureFormatDateTimeBuffer( char szTempBuffer[TEMP_BUFFER_SIZE],
 {
     const int ms = OGR_GET_MS(fSecond);
     if( ms != 0 )
-        CPLsnprintf( szTempBuffer, TEMP_BUFFER_SIZE,
-                  "%04d/%02d/%02d %02d:%02d:%06.3f",
-                  nYear,
-                  nMonth,
-                  nDay,
-                  nHour,
-                  nMinute,
-                  fSecond );
+    {
+        if ( ms == 1000 ) 
+        {
+            /* If the milliseconds are 1000 then we need to round the second up, but to get the correct date and time
+             * we need to normalize the date */
+            
+            struct tm t;
+            t.tm_year = nYear - 1900;
+            t.tm_mon = nMonth - 1;
+            t.tm_mday = nDay;
+            t.tm_hour = nHour;
+            t.tm_min = nMinute;
+            t.tm_sec = static_cast<int>(fSecond) + 1;
+            t.tm_isdst = 0;
+            
+            time_t when = mktime(&t);
```

Comment:
Instead of mktime() / localtime(), you could probably use the CPLYMDHMSToUnixTime() / CPLUnixTimeToYMDHMS() functions from cpl_time.h that don't suffer the tread-safety issue of localtime()